ValkeyProxy.md

```diff
@@ -0,0 +1,105 @@
+        c. In cluster mode, if data migration occurs, the client must handle it.
+
+        So if the client is only connected to the Proxy instead of server node, then as long as the client is not connected inSentinel mode, then when the client sends a command, it does not need to pay too much attention to the underlying connection mode. The Proxy will handle the execution of different modes for the client.
+
+3.	Support multi-database mode even server runs as cluster mode.
+
+4.	By using Proxy, clients have no perception for server failover, slot migration, adding/removing server nodes, proxy nodes and other exceptions.
+
+        Through Proxy, we can avoid the client being directly connected to the underlying server stage. All client connections, reading and writing request will only be aware of the Proxy, so the client does not need the underlying Valkey server or 
+how data is storaged. 
+
+        Therefore, no matter whether failover happens, re-electing the primary node, resharding after adding a server node, or even adding a Proxy node, clients don't need to worry about it. 
+
+        It greatly simplifies the programming complexity of the client, allowing the client to focus more on business logic without caring too much about changes in the underlying architecture.
+
+5.      Performance optimization
+
+        We can use Proxy to optimize in different ways for different clients. For example, we can support Pipeline technology on the Proxy side to improve the overall system response efficiency.
+
+        For some time-consuming or high CPU cost commands, we can make statistics in Proxy side, and for some results, such as min/max/avg/p95/p99 from memtier_benchmark or valkey_benchmark, clients can get timely statistical information.
+
+
+6.	Support audit log 
+
+        Audit log is a log that records the operations of clients accessing Proxy and Valkey server, and can provide storage,
+query, and analysis functions. With the audit log function, Proxy will automatically log the read and write requests through 
+the proxy and standardize various information such as system security events, user access records, system operation logs, and system operation status in the information system. 
+
+        Cloud providers or system administrators can utilize this information to perform rich log statistical summary and 
+correlation analysis functions to achieve a comprehensive audit of information system logs.
+
+7.      Provide reading/writing separation features
+
+        In production environments, most scenarios involve more reading requests and less writing requests. By enabling
+read-write separation feature in the Proxy, we can let the primary node focus on processing write operations and let the 
+replica handle read operations. It is very suitable for scenarios with relatively large read operations and can reduce the 
+pressure on the primary.
```

Comment:
Ooh, this can go horribly wrong. We talked with a customer that was doing caching and did read-write splits, they were having cache misses on replicas but the writes were going to the primary, but the replica was slow at consuming so it would miss again and trigger another write. I think generally read/write splitting is a big anti-pattern for online transactional workloads like caching or message queues. it makes more sense for analytic workloads, but I don't Valkey really serves much analytics today. Maybe that will change

Reply:
The reason we want to implement this feature in ValkeyProxy is: 
1. This feature (read-write separation) could be enabled or disabled through config parameter.  And by integrating with ACL, this feature only applies to some specific users
2. For those data with a relatively low update frequency, read misses due to differences in primary and replica data, which in turn lead to multiple writes, are acceptable in some scenarios.
3. If data consistency guarantees are a must following the reason 2 case, synchronous between the master and replica can be considered.